python/lsst/jointcal/plot_chi2.py

```diff
@@ -0,0 +1,262 @@
+# This file is part of jointcal.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+"""
+Extract chi2 and degrees of freedom values logged by one or more jointcal runs,
+print warnings about oddities, and make plots.
+"""
+
+import argparse
+import dataclasses
+import itertools
+import os.path
+import re
+
+import numpy as np
+import matplotlib
+matplotlib.use("Agg")  # noqa: E402
+import matplotlib.pyplot as plt
+import seaborn as sns
+sns.set_style("ticks", {"legend.frameon": True})
+sns.set_context("talk")
+
+
+@dataclasses.dataclass
+class Chi2Data:
+    """Store the chi2 values read in from a jointcal log file.
+    """
+    kind: list()
+    raw: np.ndarray
+    ndof: np.ndarray
+    reduced: np.ndarray
+    init_count: int = dataclasses.field(init=False)
+
+    def __post_init__(self):
+        # ensure the array types are correct
+        self.raw = np.array(self.raw, dtype=np.float64)
+        self.ndof = np.array(self.ndof, dtype=np.int)
+        self.reduced = np.array(self.reduced, dtype=np.float64)
+        self.init_count = self._find_init()
+
+    def _find_init(self):
+        """Return the index of the first "fit step", after initialization.
+
+        NOTE
+        ----
+        There are never more than ~25 items in the list, so search optimization
+        is not worth the trouble.
+        """
+        # Logs pre-DM-25779
+        if "Fit prepared" in self.kind:
+            return self.kind.index("Fit prepared") + 1
+        # Logs post-DM-25779
+        elif "Fit iteration 0" in self.kind:
+            return self.kind.index("Fit iteration 0")
+        else:
+            raise RuntimeError(f"Cannot find end of initialization sequence in {self.kind}")
+
+
+class LogParser:
+    """Parse a jointcal logfile to extract chi2 values and plot them.
+
+    Call the instance with the path to a file to check it for anamolous chi2
+    and output plots to your current directory.
+
+    Parameters
+    ----------
+    plot : `bool`
+        Make plots for each file (saved to the current working directory)?
+    verbose : `bool`
+        Print updates during processing.
```

Comment:
Nitpick, but why does one docstring end with a question mark, and the other a period?